### PR TITLE
feat(rules): add tally/ruby/prefer-secret-mounts-for-build-credentials

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

Adds [`tally/ruby/prefer-secret-mounts-for-build-credentials`](https://tally.wharflab.com/rules/tally/ruby/prefer-secret-mounts-for-build-credentials/) — Section 4.4 of [design-docs/43-ruby-on-docker.md](../blob/main/design-docs/43-ruby-on-docker.md).

Constructive companion to [`tally/ruby/secrets-in-arg-or-env`](https://tally.wharflab.com/rules/tally/ruby/secrets-in-arg-or-env/) (PR #678). Fires at info severity when ARG/ENV declares a Ruby-ecosystem build-time credential, pointing the user at BuildKit secret mounts.

- **Coverage:** `BUNDLE_GITHUB__COM` / `BUNDLE_BITBUCKET__ORG` / `BUNDLE_GITLAB__COM` (well-known Bundler hosts), `BUNDLE_<HOST>__<TLD>` (generic Bundler host pattern), `GEM_HOST_API_KEY` (RubyGems push), `NPM_TOKEN` / `YARN_AUTH_TOKEN` (used during `assets:precompile` for private npm packages).
- **Detection:** triggers on both ARG and ENV, including meta-ARG (before any FROM).
- **Suppressions:** dev/test/ci stages, non-Ruby, Windows. Non-credential `BUNDLE_*` vars (e.g. `BUNDLE_PATH`) don't match the host pattern.
- **Auto-fix:** `FixSuggestion` (no-edit). Description names the canonical `RUN --mount=type=secret,id=<id>,env=<VAR>` form with a suggested `secret_id` (case-lowered env-var name).

Corpus uptake: 1 of 196 Dockerfiles use `RUN --mount=type=secret` (`basecamp/fizzy/saas` is the textbook example).

## Test plan

- [x] `make build`
- [x] `make test` — ruby package and integration suite pass
- [x] `make lint` — 0 issues
- [x] Lint fixture: `internal/integration/fixtures/lint/ruby-prefer-secret-mounts-for-build-credentials/`

### CODE REVIEW

**Comment by tinovyatkin** [comment]
@coderabbitai review

**Comment by tinovyatkin** [comment]
@codex review

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+			switch c := cmd.(type) {
+			case *instructions.ArgCommand:
+				if v := r.checkArg(input, c, meta); v != nil {
+					violations = append(violations, *v)
+				}
+			case *instructions.EnvCommand:
+				violations = append(violations, r.checkEnv(input, c, meta)...)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go]: Both addressed in 9bd00a86:

1. **Nil-Facts guard**: Added `if input.Facts == nil { continue }` check in the per-stage loop and `dockerfileHasRubyStage` helper. The rule now no-ops when Facts isn't populated (which shouldn't happen in practice but is defensive against future test harness changes).
2. **Slice return for ARG**: `checkArg` now returns `[]rules.Violation` so a single `ARG VAR1 VAR2` declaration with multiple credential keys yields one violation per key. Common emit logic factored into `emitViolations` to keep the dupl analyzer satisfied.

Added tests for both: nil-Facts is exercised through normal lint plumbing; multi-var ARG case covered explicitly by `ARG_with_multiple_credential_vars_yields_one_violation_per_var`.

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+		}
+		loc := rules.NewLocationFromRanges(input.File, cmd.Location())
+		v := rules.NewViolation(loc, meta.Code, meta.Description, meta.DefaultSeverity).
+			WithDocURL(meta.DocURL).
+			WithDetail(preferSecretMountsDetail(arg.Key, secretID, strings.ToUpper(command.Arg))).
+			WithSuggestedFix(buildSecretMountFix(arg.Key, secretID, meta.FixPriority))
+		return &v
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go:155]: Done in 9bd00a86 — `checkArg` now returns `[]rules.Violation` and collects all matching keys (e.g. `ARG BUNDLE_GITHUB__COM BUNDLE_GITLAB__COM` produces 2 violations). Test case added.

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+		if v := r.checkArg(input, &arg, meta); v != nil {
+			violations = append(violations, *v)
+		}
+	}
+
+	for stageIdx, stage := range input.Stages {
+		if stagename.LooksLikeDev(stage.Name) {
+			continue
+		}
+		sf := input.Facts.Stage(stageIdx)
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go:89]: Addressed in 9bd00a86 — added `stageLooksLikeRuby` gate before per-stage scanning. A Node stage with `ENV NPM_TOKEN=...` no longer fires this Ruby rule. Test case `Node.js stage with NPM_TOKEN does not trigger` covers it.

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+func (r *PreferSecretMountsForBuildCredentialsRule) Check(input rules.LintInput) []rules.Violation {
+	meta := r.Metadata()
+
+	var violations []rules.Violation
+
+	for _, arg := range input.MetaArgs {
+		if v := r.checkArg(input, &arg, meta); v != nil {
+			violations = append(violations, *v)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go]: Addressed in 9bd00a86 — added `dockerfileHasRubyStage` helper. Meta-ARGs are only checked when at least one stage in the Dockerfile is Ruby-shaped. Test case `meta-ARG without any Ruby stage does not trigger` covers it.

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+			case *instructions.ArgCommand:
+				if v := r.checkArg(input, c, meta); v != nil {
+					violations = append(violations, *v)
+				}
+			case *instructions.EnvCommand:
+				violations = append(violations, r.checkEnv(input, c, meta)...)
+			}
+		}
+	}
+	return violations
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go:111]: Done in 9bd00a86 — same `stageLooksLikeRuby` gate as the other rules in this package. Updated docs to call out the gate explicitly. Three test cases cover it: Node stage with `NPM_TOKEN`, Node stage with `YARN_AUTH_TOKEN`, meta-ARG without any Ruby stage.

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+	return strings.Contains(rest, "__")
+}
+
+func preferSecretMountsDetail(envKey, secretID, instruction string) string {
+	return "`" + envKey + "` is a build-time credential; declaring it via " + instruction + " bakes the " +
+		"value into image history (`docker history --no-trunc <image>`) and into the build cache key " +
+		"data. Pass it through a BuildKit secret mount instead — the secret exists for the duration of " +
+		"the RUN it's mounted into and never enters image content or cache key data: " +
+		"`RUN --mount=type=secret,id=" + secretID + ",env=" + envKey + " bundle install`."
+}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go:296]: Done in 9bd00a86 — the example command in the detail message is now `<command>` instead of `bundle install`, so the suggestion reads correctly for `NPM_TOKEN`/`YARN_AUTH_TOKEN` (and is still actionable for Bundler-host vars: users substitute their own command). Snapshot updated.

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+// `__` (double underscore) for `.` (dot), so a real Bundler key has at
+// least one `__` separator.
+func isBundlerHostCredentialKey(key string) bool {
+	if !strings.HasPrefix(key, "BUNDLE_") {
+		return false
+	}
+	// Must have at least one `__` after the BUNDLE_ prefix to look like
+	// a host (dot-separated) credential.
+	rest := strings.TrimPrefix(key, "BUNDLE_")
+	return strings.Contains(rest, "__")
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go]: Good catch. Addressed in `024f8f07`:

The matcher now requires the trailing `__<TLD>` segment to be a recognized public-suffix TLD (`com`, `org`, `net`, `io`, `dev`, `co`, `ai`). `BUNDLE_LOCAL__RACK` (Bundler's `local.rack` config) no longer matches; `BUNDLE_GEMFURY__IO` still does.

Two new test cases lock this in: `BUNDLE_LOCAL__RACK is a config key (local.rack), not a credential` (suppressed) and `BUNDLE_GEMFURY__IO triggers (.io TLD)` (flagged).

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+func dockerfileHasRubyStage(input rules.LintInput) bool {
+	if input.Facts == nil {
+		return false
+	}
+	for stageIdx, stage := range input.Stages {
+		sf := input.Facts.Stage(stageIdx)
+		if sf == nil {
+			continue
+		}
+		if stageLooksLikeRuby(input.Semantic, stageIdx, stage, sf) {
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go:135]: Addressed in `024f8f07`. `dockerfileHasRubyStage` now applies the same dev/test/Windows skip filters used in the per-stage scan. A Dockerfile whose only Ruby stage is named `dev` no longer triggers meta-ARG findings.

Test case `meta-ARG with only dev Ruby stage does not trigger` covers this.

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+	}
+	rest := strings.TrimPrefix(key, "BUNDLE_")
+	if !strings.Contains(rest, "__") {
+		return false
+	}
+	// The trailing `__<TLD>` segment must be a recognized public-suffix
+	// TLD. Take the last `__`-separated token.
+	idx := strings.LastIndex(rest, "__")
+	tld := rest[idx+2:]
+	return bundlerHostCredentialTLDs[tld]
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go]: Right, the allowlist was the wrong shape — Bundler's host-credential surface is open-ended. Switched to a denylist in `bfd31c7a`:

`isBundlerHostCredentialKey` now treats any `BUNDLE_<TOPIC>__<rest>` as a host credential **except** the known non-host config namespaces (`BUNDLE_LOCAL__<gem>`, `BUNDLE_BUILD__<gem>`). That preserves the original false-positive fix (`BUNDLE_LOCAL__RACK`) without dropping coverage.

Three new tests:
- `BUNDLE_BUILD__NOKOGIRI` (per-gem build flag) — suppressed
- `BUNDLE_GEMS__ACME__CO__UK` (compound TLD) — flagged
- `BUNDLE_PRIVATE__HOST__DE` (German TLD) — flagged

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+func isBundlerHostCredentialKey(key string) bool {
+	if !strings.HasPrefix(key, "BUNDLE_") {
+		return false
+	}
+	rest := strings.TrimPrefix(key, "BUNDLE_")
+	first, _, hasSep := strings.Cut(rest, "__")
+	if !hasSep {
+		return false
+	}
+	return !bundlerNonHostConfigPrefixes[first]
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go]: Addressed in `7af4c0b4` — extended the denylist to cover all known Bundler `__`-encoded config namespaces:

- `LOCAL` (`local.<gem>`) — gem development paths
- `BUILD` (`build.<gem>`) — per-gem build flags
- `PATH` (`path.system` and similar) — path config
- `GEM` (`gem.coc`/`gem.mit`/`gem.test`/etc.) — gem-template flags
- `CACHE` (`cache.all`/`cache.path`) — cache config
- `DISABLE` (`disable.foo`) — feature disable flags

Tests cover one example per namespace.

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go]: You're right that the denylist was unsustainable — every iteration revealed another Bundler config namespace I'd missed. Switched to positive identification in `33226171`:

The trailing `__<token>` segment must look like a TLD. Acceptance:
- token is in a curated common-TLD list (com/org/net/io/dev/co/ai/edu/gov/info/biz/app), OR
- token is a 2-letter all-alphabetic ccTLD (`de`/`uk`/`jp`/`br`/…).

This positively identifies host credentials and naturally excludes the entire family of dotted Bundler config namespaces without needing to enumerate them. `BUNDLE_IGNORE_MESSAGES__HTTPARTY` is suppressed (HTTPARTY isn't a TLD); `BUNDLE_PRIVATE__HOST__DE` is flagged (DE is a ccTLD).

Tests cover: ignore_messages suppression, country-code TLD trigger, compound TLD trigger, and the previously-fixed local/build/path/gem/cache/disable cases.

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+	if !strings.Contains(rest, "__") {
+		return false
+	}
+	idx := strings.LastIndex(rest, "__")
+	tld := strings.ToLower(rest[idx+2:])
+	if commonHostCredentialTLDs[tld] {
+		return true
+	}
+	// 2-letter all-alphabetic ccTLD (e.g. `de`, `uk`, `jp`, `br`).
+	if len(tld) == 2 && isAllAlpha(tld) {
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go]: You raised these two points together and they pull in opposite directions — accept more TLDs vs suppress more 2-letter false positives. Resolving by combining both gates in `3e8c673d`:

**Gate 1**: Explicit denylist of Bundler-known dotted-config namespaces (LOCAL/BUILD/PATH/GEM/CACHE/DISABLE/IGNORE_MESSAGES) — these are configuration even if the trailing token looks like a TLD.

**Gate 2**: The trailing token must be shaped like a TLD: 2–6 chars, all ASCII letters. Rejects gem names (typically longer or contain digits/underscores) but accepts the full long tail of legitimate TLDs (`xyz`, `museum`, country codes, etc.).

A `BUNDLE_X__Y` key is treated as a host credential ONLY when both gates pass. So:
- `BUNDLE_LOCAL__AI` → suppressed (gate 1 fails: LOCAL is config)
- `BUNDLE_GEMS__ACME__MUSEUM` → flagged (gate 1 passes; MUSEUM is TLD-shaped, 6 chars)
- `BUNDLE_PRIVATE__HOST__XYZ` → flagged (gate 1 passes; XYZ is TLD-shaped)

Tests cover both edge cases the comments highlighted.

File: internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+		return true
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/ruby/prefer_secret_mounts_for_build_credentials.go]: You raised these two points together and they pull in opposite directions — accept more TLDs vs suppress more 2-letter false positives. Resolving by combining both gates in `3e8c673d`:

**Gate 1**: Explicit denylist of Bundler-known dotted-config namespaces (LOCAL/BUILD/PATH/GEM/CACHE/DISABLE/IGNORE_MESSAGES) — these are configuration even if the trailing token looks like a TLD.

**Gate 2**: The trailing token must be shaped like a TLD: 2–6 chars, all ASCII letters. Rejects gem names (typically longer or contain digits/underscores) but accepts the full long tail of legitimate TLDs (`xyz`, `museum`, country codes, etc.).

A `BUNDLE_X__Y` key is treated as a host credential ONLY when both gates pass. So:
- `BUNDLE_LOCAL__AI` → suppressed (gate 1 fails: LOCAL is config)
- `BUNDLE_GEMS__ACME__MUSEUM` → flagged (gate 1 passes; MUSEUM is TLD-shaped, 6 chars)
- `BUNDLE_PRIVATE__HOST__XYZ` → flagged (gate 1 passes; XYZ is TLD-shaped)

Tests cover both edge cases the comments highlighted.
